packages/grafana-data/src/field/fieldState.ts

```diff
@@ -62,11 +62,32 @@ export function cacheFieldDisplayNames(frames: DataFrame[]) {
 /**
  *
  * moves each field's config.custom.hideFrom to field.state.hideFrom
- * and mutates orgiginal field.config.custom.hideFrom to one with explicit overrides only, (without the ad-hoc stateful __system override from legend toggle)
+ * and sets field.config.custom.hideFrom to one with explicit overrides only, (without the ad-hoc stateful __system override from legend toggle)
  */
 export function decoupleHideFromState(frames: DataFrame[], fieldConfig: FieldConfigSource) {
-  frames.forEach((frame) => {
-    frame.fields.forEach((field) => {
+  return frames.map((frame) => {
+    const frameCopy: DataFrame = { ...frame };
+
+    frameCopy.fields = frame.fields.map((field) => {
+      const fieldCopy: Field = {
+        ...field,
+        state: {
+          ...field.state,
+          hideFrom: {
+            ...(field.state?.hideFrom ?? { legend: false, tooltip: false, viz: false }),
+          },
+        },
+        config: {
+          ...field.config,
+          custom: {
+            ...field.config.custom,
+            hideFrom: {
+              ...field.config.custom?.hideFrom,
+            },
+          },
+        },
+      };
```

Comment:
this removes the mutation, allowing the before/after config diff to invalidate and re-init uPlot